File: src/php/AbstractOasParser.php
<?php

namespace TuEmpresa\LaravelAiApiIntegration\Infrastructure\Services\OasParser;

use TuEmpresa\LaravelAiApiIntegration\Domain\Interfaces\OasParserInterface;
use TuEmpresa\LaravelAiApiIntegration\Domain\Entities\ApiSpecification;
use Symfony\Component\Yaml\Yaml;
use TuEmpresa\LaravelAiApiIntegration\Domain\Exceptions\OasParserException;

abstract class AbstractOasParser implements OasParserInterface {
    protected array $data = [];

    protected function loadData(string $oasFilePath): void {
        try {
            $content = file_get_contents($oasFilePath);

            if ($content === false) {
                throw new OasParserException("Unable to read file: $oasFilePath");
            }

            $extension = pathinfo($oasFilePath, PATHINFO_EXTENSION);
            
            if ($extension === 'json') {
                $this->data = json_decode($content, true, 512, JSON_THROW_ON_ERROR);
            } elseif ($extension === 'yaml' || $extension === 'yml') {
                $this->data = Yaml::parse($content);
            } else {
                throw new OasParserException("Unsupported file format: $extension");
            }

            if (empty($this->data)) {
                throw new OasParserException("Failed to parse file content as " . strtoupper($extension));
            }

            if (!isset($this->data['openapi'])) {
                throw new OasParserException("Invalid OpenAPI specification: 'openapi' version not found.");
            }
        } catch (\JsonException $e) {
            throw new OasParserException("Failed to parse JSON: " . $e->getMessage());
        } catch (\Symfony\Component\Yaml\Exception\ParseException $e) {
            throw new OasParserException("Failed to parse YAML: " . $e->getMessage());
        }
    }

    abstract public function parse(string $oasFilePath = ''): ApiSpecification;

    abstract protected function getApiName(): string;
    abstract protected function getBaseUrl(): string;
    abstract protected function getVersion(): string;

    protected function generateOperationId(string $method, string $path): string {
        $parts = explode('/', ltrim($path, '/'));
        $parts = array_filter($parts, fn($part) => !preg_match('/^ {.*}$/', $part));
        $operationId = strtolower($method) . ucfirst(implode('', array_map('ucfirst', $parts)));
        return lcfirst($operationId);
    }
}

File: src/infrastructure/services/oas_parser/abstract_oas_parser.py
import yaml
import json
from pathlib import Path
from src.domain.interfaces.oas_parser_interface import OasParserInterface
from src.domain.entities.api_specification import ApiSpecification
from src.domain.exceptions.oas_parser_exception import OasParserException
from src.domain.value_objects.yml_obj import YmlObj

class AbstractOasParser(OasParserInterface):
    def __init__(self):
        self.data = YmlObj({})

    def load_data(self, oas_file_path: str) -> None:
        try:
            path = Path(oas_file_path)
            content = path.read_text()

            if path.suffix == '.json':
                self.data = YmlObj(json.loads(content))
            elif path.suffix in ['.yaml', '.yml']:
                self.data = YmlObj(yaml.safe_load(content))
            else:
                raise OasParserException(f"Unsupported file format: {path.suffix}")

            if not self.data:
                raise OasParserException(f"Failed to parse file content as {path.suffix.upper()}")

            if 'openapi' not in self.data:
                raise OasParserException("Invalid OpenAPI specification: 'openapi' version not found.")
        except json.JSONDecodeError as e:
            raise OasParserException(f"Failed to parse JSON: {str(e)}")
        except yaml.YAMLError as e:
            raise OasParserException(f"Failed to parse YAML: {str(e)}")
        except Exception as e:
            raise OasParserException(f"Error loading OAS file: {str(e)}")

    @staticmethod
    def generate_operation_id(method: str, path: str) -> str:
        parts = [part for part in path.lstrip('/').split('/') if not part.startswith('{')]
        operation_id = method.lower() + ''.join(part.capitalize() for part in parts)
        return operation_id[0].lower() + operation_id[1:]
